enhancements/single-node/single-node-openshift-with-workers.md

```diff
@@ -0,0 +1,482 @@
+
+- `SNO` = Single control-plane node cluster
+
+|Scenario                                |Platform|D1W|D2W|CIT|PIT|CTP|PTP|Is a load balancer currently needed?             |Will a load balancer be needed after proposal?|
+|----------------------------------------|--------|---|---|---|---|---|---|-------------------------------------------------|----------------------------------------------|
+|None platform SNO no workers            |None    |0  |0  |SR |SR |W  |M  |No                                               |No                                            |
+|Cloud platform SNO no workers           |Cloud   |0  |0  |SR |SR |W  |M  |No                                               |No                                            |
+|None platform SNO day1 single worker    |None    |1  |1  |-  |SR |-  |M  |Yes, ingress floats                              |No, ingress pinned to master pool             |
+|Cloud platform SNO day1 single worker   |Cloud   |1  |1  |SR |HA |W  |W  |Yes - provided, ingress floats                   |Yes - provided, there are two ingress replicas|
+|None platform SNO day1 multiple workers |None    |2+ |2+ |-  |SR |-  |M  |Yes, ingress is on two different nodes           |No, one ingress and is pinned to master pool  |
+|Cloud platform SNO day1 multiple workers|Cloud   |2+ |2+ |HA |HA |W  |W  |Yes - provided, ingress is on two different nodes|Yes - provided, there are two ingress replicas|
+|None platform SNO only day2 workers     |None    |0  |1+ |SR |SR |W  |M  |Yes, ingress floats                              |No, ingress pinned to master pool             |
+|Cloud platform SNO only day2 workers    |Cloud   |0  |1+ |SR |SR |W  |M  |Yes - provided, ingress floats                   |Yes - provided, ingress floats                |
+
+Some notes about this table and the decisions made in it:
+
+The table assumes that the "Infrastrcture Topology" value is read only. That
+means an optimal value for this parameter must be determined during
+installation. You can see the consequences of that, for example, in the last
+row of the table - even though we add workers in day 2 to the cluster, the
+"Infrastrcture Topology" stays "SingleReplica" even though that cluster could
+benefit from the additional worker nodes for the purposes of highly-available
+infrastructure (as it's running in the cloud, so it has a load-balancer). In
+the future it may be worth revisiting making this value configurable.
+
+The "None platform SNO day1 single/multiple worker" scenario rows aren't really
+possible today as there is no real way to install this kind of cluster. That's
+why their CIT / CTP have been left empty.
+
+It's proposed that the only scenarios in which the "Infrastrcture Topology"
+parameter will be "HighlyAvailable" are when installing a single control-plane
+node cluster in the cloud with one or more day-1 workers. Cloud deployments
+have a load balancer so they can benefit from having highly-available ingress.
```

Comment:
We don't currently offer formal support for single-node deployments in the cloud. Is that expected to change as part of this work?

If the reason for using the HA setting is the presence of a load balancer, would we want to make it possible to configure the infrastructure topology for HA on metal (or platform none) deployments if there is also a load balancer there?

Reply:
> We don't currently offer formal support for single-node deployments in the cloud. Is that expected to change as part of this work?
nope. Cloud support is out of scope. But we did want to address the flow regardless.

Regrading HA for infrastructure topology - I personally think that it should always be SingleReplica. I think if customers need any kind of HA, they should move to 3nodes + workers cluster. Also, I dont think we should differentiate between on prem and cloud envs unless it's a real must

Reply:
> Regrading HA for infrastructure topology - I personally think that it should always be SingleReplica. I think if customers need any kind of HA, they should move to 3nodes + workers cluster. Also, I dont think we should differentiate between on prem and cloud envs unless it's a real must

Being consistent is going to make it easier to reason about the system, so I support that as well, especially if we're still considering formal cloud support out of scope.

Reply:
I addressed this point in the third bullet in the "Alternatives section".

Some caveats -

- Setting SingleReplica in those scenarios would mean Control plane topology == infrastructure topology under all circumstances, these two parameters would always be identical unless we one day allow users to choose / change them.

- Making this consistent across the types of single node deployments means we're losing consistency with regular HA OCP deployments

- Cloud deployments would lose the HA ingress benefits they can get from the load balancer

Anyway I don't think the decision here is too critical, I don't mind either way. I'll update the enhancement to go with SingleReplica everywhere if you confirm you're okay with the above caveats

Reply:
The control plane topology can also be "external", in the Hypershift case. So the infrastructure and control plane fields wouldn't always be the same.